useful_types/experimental.py

```diff
@@ -22,8 +34,33 @@
 # Objects suitable to be passed to sys.settrace, threading.settrace, and similar
 TraceFunction: TypeAlias = Callable[[FrameType, str, Any], Union["TraceFunction", None]]
 
-# Might not work as expected for pyright, see
-#   https://github.com/python/typeshed/pull/9362
-#   https://github.com/microsoft/pyright/issues/4339
-class DataclassInstance(Protocol):
-    __dataclass_fields__: ClassVar[dict[str, Field[Any]]]
+if TYPE_CHECKING:
+    # Might not work as expected for pyright, see
+    #   https://github.com/python/typeshed/pull/9362
+    #   https://github.com/microsoft/pyright/issues/4339
+    #
+    # The type ignore is to workaround a mypy complaint:
+    # Final class useful_types.experimental.DataclassLike has abstract attributes "__dataclass_fields__"
+    @final
+    @runtime_checkable
+    class DataclassLike(Protocol):  # type: ignore[misc]
```

Comment:
Mypy needs this `type: ignore`, but pyright is erroring out saying that it's unused. We could employ a similar strategy to the one we use over at typeshed, where we tell pyright to ignore all `type: ignore` comments, and only use `pyright: ignore` comments for pyright.

Reply:
Good point, we're going to need separate config for `static_tests` and `useful_types`.